File: leetcode/0897-increasing-order-search-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* increasingBST(TreeNode* root) {
        if (!root)
            return root;
        
        TreeNode *ansHead = NULL, *ansTail = NULL;
        stack<TreeNode*> s;
        TreeNode *store = root;
        
        while (!s.empty() || store != NULL) {
            if (store != NULL) {
                s.push(store);
                store = store->left;
            } else {
                TreeNode *top = s.top();
                s.pop();
                if (ansHead == NULL) {
                    ansHead = top;
                    ansTail = top;
                } else {
                    ansTail->right = top;
                    ansTail = top;
                }
                
                top->left = NULL;
                
                store = top->right;
            }
        }
        
        return ansHead;
    }
};

File: leetcode/0897-increasing-order-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def increasingBST(self, root: TreeNode) -> TreeNode:
        if root == None:
            return root
        
        ansHead, ansTail = None, None
        stack = []
        store = root
        
        while len(stack) != 0 or store != None:
            if store != None:
                stack.append(store)
                store = store.left
            else:
                top = stack.pop()
                
                if ansHead == None:
                    ansHead = top
                    ansTail = top
                else:
                    ansTail.right = top
                    ansTail = top
                    
                top.left = None
                
                store = top.right
        
        return ansHead